src/v/storage/mvlog/batch_collector.cc

```diff
@@ -0,0 +1,74 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "storage/mvlog/batch_collector.h"
+
+#include "storage/mvlog/errc.h"
+#include "storage/mvlog/reader_outcome.h"
+
+namespace storage::experimental::mvlog {
+
+bool batch_collector::should_skip(
+  const model::record_batch_header& batch_hdr) const noexcept {
+    if (batch_hdr.last_offset() < cfg_.start_offset) {
+        return true;
+    }
+    if (batch_hdr.max_timestamp < cfg_.first_timestamp) {
+        return true;
+    }
+    if (
+      cfg_.type_filter.has_value()
+      && cfg_.type_filter.value() != batch_hdr.type) {
+        return true;
+    }
+    return false;
+}
+
+result<reader_outcome, errc> batch_collector::add_batch(
+  model::record_batch_header batch_hdr, iobuf records) noexcept {
+    if (unlikely(batch_hdr.base_offset <= last_offset_)) {
+        // We're expecting to read through monotonically increasing batches!
+        return errc::broken_data_invariant;
+    }
+    // TODO: add considerations around `strict_max_bytes`, or handle bytes
+    // limiting in a lower layer.
+
+    last_offset_ = batch_hdr.last_offset();
+    if (batch_hdr.base_offset() > cfg_.max_offset) {
+        return reader_outcome::stop;
+    }
+
+    if (should_skip(batch_hdr)) {
+        return reader_outcome::skip;
+    }
+    // TODO: add ghost batch building here.
+
+    cur_buffer_size_ += batch_hdr.size_bytes;
+    batch_hdr.ctx.term = cur_term_;
```

Comment:
got it so the term isn't stored in the batches on disk, we're going to mix that in as we go along from some other source?

Reply:
Right, at least in my local branch, the term is going to be stored as a part of the entry body envelope for record batches. For now I'm just reusing the disk serialization for record batches we have today in segment appender, rather than using the envelope, since we already have an envelope to wrap the record batch guts.